pallets/network-membership/src/lib.rs

```diff
@@ -0,0 +1,422 @@
+		/// Rejects request if the member is added to the blacklist
+		MembershipRequestRejected,
+		/// Membership expired
+		MembershipExpired,
+		/// Max members limit exceeded
+		MaxMembersExceededForTheBlock,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			if n > T::BlockNumber::zero() {
+				Self::renew_or_expire_memberships(n)
+			} else {
+				Weight::zero()
+			}
+		}
+	}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub members: BTreeMap<T::AccountId, MemberData<T::BlockNumber>>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self { members: Default::default() }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			for (member, member_data) in &self.members {
+				Members::<T>::insert(member, member_data);
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Add an author. Only root or council origin can perform this
+		/// action.
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::nominate())]
+
+		pub fn nominate(
+			origin: OriginFor<T>,
+			member: CordAccountOf<T>,
+			expires: bool,
+		) -> DispatchResult {
+			T::NetworkMembershipOrigin::ensure_origin(origin)?;
+
+			// Check if member already exist it should throw error
+			// 'MembershipAlreadyAcquired'
+			ensure!(!<Members<T>>::contains_key(&member), Error::<T>::MembershipAlreadyAcquired);
+
+			if expires {
+				Self::insert_membership_and_schedule_expiry(member.clone())
+			} else {
+				let expire_on = T::BlockNumber::zero();
+				Members::<T>::insert(&member, MemberData { expire_on });
+			}
+
+			// the member has just been created, increment its provider
+			let _ = frame_system::Pallet::<T>::inc_providers(&member);
+
+			Self::deposit_event(Event::MembershipAcquired { member });
+
+			Ok(())
+		}
+
+		/// Renew authorship. Only root or council orgin can perform this
+		/// action.
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::renew())]
+
+		pub fn renew(origin: OriginFor<T>, member: CordAccountOf<T>) -> DispatchResult {
+			T::NetworkMembershipOrigin::ensure_origin(origin)?;
+
+			// Check if membership renewal request already exist it should throw error
+			// 'MembershipRenewalAlreadyRequested'
+			ensure!(
+				!<MembershipsRenewsOn<T>>::contains_key(&member),
+				Error::<T>::MembershipRenewalAlreadyRequested
+			);
+
+			MembershipsRenewsOn::<T>::insert(&member, ());
+
+			Self::deposit_event(Event::MembershipRenewalRequested { member });
+
+			Ok(())
+		}
+
+		/// Revoke a membership. Only root or council orgin can perform this
+		/// action.
+		#[pallet::call_index(2)]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::revoke())]
+		pub fn revoke(origin: OriginFor<T>, member: CordAccountOf<T>) -> DispatchResult {
```

Comment:
just a thought: should we keep an optional 'reason' string as an (256 bytes max) parameter. This needn't be stored anywhere, but be part of extrinsic so there is a reference (works good only if there are different reason for revoking a membership... like "inactive", "no show in voting for N months" etc.

Reply:
That's good to have. An enum-based approach would be nice compared to vec.